wifi: shell: refactor wifi_shell.c

ndrs-pst opened the pull request:
- introduces a local variable `sh` to store `context.sh` for use with the `shell_fprintf` function. By doing so, we avoid the repeated dereferencing of `context.sh`, thereby reducing the code footprint.

- Moved `scan_result` from a standalone variable into the `context` struct to enhance code optimization.

- Replaced direct `shell_fprintf` calls with `PR`, `PR_ERROR`, `PR_INFO` and `PR_WARNING` macros.

ndrs-pst committed:
wifi: shell: introduce local 'sh' variable

This commit introduces a local variable 'sh' to store 'context.sh' for
use with the 'shell_fprintf' function.

By doing so, we avoid the repeated dereferencing of 'context.sh',
thereby reducing the code footprint.

Signed-off-by: Pisit Sawangvonganan <pisit@ndrsolution.com>

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -184,11 +184,12 @@ static void handle_wifi_raw_scan_result(struct net_mgmt_event_callback *cb)
 	int rssi;
```
How is this reducing code footprint, you have just an extra variable in the code which is probably optimized out by the compiler although I did not check this.

ndrs-pst replied:
Hi @jukkar, FWIW, I built **"samples/net/wifi"** on the **"disco_l475_iot1"** board, using various optimizations detailed in the table below.

| OPTIMIZATIONS  | FLASH SIZE (BEFORE) | FLASH SIZE (AFTER) | REDUCE (AFTER - BEFORE) |
| ------------- | --------------------------- | ---------------------------------- | ---- |
| SIZE_OPTIMIZATIONS  | 175,248 B | 175,176 B | -72 B |
| SPEED_OPTIMIZATIONS  | 202,552 B | 202,472 B | -80 B |
| DEBUG_OPTIMIZATIONS  | 209,396 B | 209,372 B | -24 B |

---
When `SIZE_OPTIMIZATIONS=y`, the code footprint of `wifi_shell.c` is reduced by **72 bytes**, 
from 9,084 bytes to 9,012 bytes.

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -62,7 +62,7 @@ static struct {
 		uint8_t all;
 	};
-} context;
+} wifi_shell_ctx;
```
For the second commit, I am not really sure how this improves clarity. The original code looks fine by me.

ndrs-pst replied:
Then I will revoke this change.

jukkar commented:
What you are proposing here are not really needed, but I would not mind if we could introduce similar print macros as what net-shell is using. See https://github.com/zephyrproject-rtos/zephyr/blob/018dbcfd6679c273842084ce34c167295bc6f354/subsys/net/lib/shell/net_shell_private.h#L11 for details.

ndrs-pst committed:
wifi: shell: move "scan_result" variable into context struct

Moved 'scan_result' from a standalone variable into the 'context' struct
to enhance code optimization.

Signed-off-by: Pisit Sawangvonganan <pisit@ndrsolution.com>

ndrs-pst committed:
wifi: shell: refactor to use PR(...) from "net_shell_private.h"

Replaced direct shell_fprintf calls with PR, PR_ERROR, PR_INFO
and PR_WARNING macros.

This change simplifies the code by using predefined macros.

Signed-off-by: Pisit Sawangvonganan <pisit@ndrsolution.com>

ndrs-pst commented:
> What you are proposing here are not really needed, but I would not mind if we could introduce similar print macros as what net-shell is using. See
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/018dbcfd6679c273842084ce34c167295bc6f354/subsys/net/lib/shell/net_shell_private.h#L11
> 
> for details.

Then I may bring `#include "net_shell_private.h"` to `wifi_shell.c` and use such macro

rlubos commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -798,43 +798,38 @@ static int cmd_wifi_status(const struct shell *sh, size_t argc, char *argv[])
+		PR("Interface Mode: %s\n", wifi_mode_txt(status.iface_mode));
+		PR("Link Mode: %s\n", wifi_link_mode_txt(status.link_mode));
+		PR("SSID: %.32s\n", status.ssid);
+		PR("BSSID: %s\n",
+		   net_sprint_ll_addr_buf(status.bssid,
+		   WIFI_MAC_ADDR_LEN, mac_string_buf,
+		   sizeof(mac_string_buf))
+		  );
```
Since we're already touching those lines I suggest to modify the alignment a bit, so that it's clear where each argument belongs. 
```suggestion
		PR("BSSID: %s\n",
		   net_sprint_ll_addr_buf(status.bssid, 
		   			  WIFI_MAC_ADDR_LEN, mac_string_buf,
		   			  sizeof(mac_string_buf)));
```

ndrs-pst replied:
The changes have been incorporated and squashed into this commit: https://github.com/zephyrproject-rtos/zephyr/pull/68185/commits/236963e6c6f4245fadb2b34c7729f3bf3dbdc230

rlubos reviewed (dismissed):
No objections from my side, I was quite surprised to see that such changes improve flash footprint, so in that case I don't mind minor refactoring.

ndrs-pst commented:
@rlubos, I have another question: is there any chance that the shell subsystem will pass `NULL` into the `'sh'` argument?

Since the `print(sh, level, fmt, ...)` macros seem to be trying to protect against this case.

jukkar commented:
> I have another question: is there any chance that the shell subsystem will pass `NULL` into the `'sh'` argument?
> 
> Since the `print(sh, level, fmt, ...)` macros seem to be trying to protect against this case.

I have never seen that happen that sh argument would be null. Lets not worry about that too much, and if it becomes a problem, we can adjust the macros.

ndrs-pst commented:
> @ndrs-pst can you rebase please?

@fabiobaltieri, No problem, I've already done it.

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -546,8 +546,8 @@ static int __wifi_args_to_params(size_t argc, char *argv[],
 			if (!found) {
 				print(sh, SHELL_ERROR,
```
Shouldn't these now be `PR_ERROR`?

ndrs-pst replied:
The print(sh, ...) macros appear to handle a NULL `sh` value, as seen in the code

https://github.com/zephyrproject-rtos/zephyr/blob/b3e67cf2a62b01fbf9f4c0eb618cd676b4dc3072/subsys/net/l2/wifi/wifi_shell.c#L79-L86

Therefore, I have left it unchanged.
However, as @jukkar mentioned, there is no likelihood that `sh` will be `NULL`.
If this is confirmed, I will then apply the change to all places where the `print(sh, ...)` macros are used and remove those macros.

krish2718 replied:
Sure, it does handle `NULL`, but better to stick with a single macro to print.

ndrs-pst replied:
Ok, so I applied this commit: https://github.com/zephyrproject-rtos/zephyr/pull/68185/commits/2fe89d67e5afab360e454b37361f8f8942e3506c

krish2718 reviewed (dismissed):
A few print's were missed, probably during rebase.

ndrs-pst committed:
wifi: shell: replace print(...) with PR(...)

Additionally, replace the local print(sh, level, fmt, ...) macros
with PR, PR_ERROR, and PR_WARNING macros.

Then remove the print(sh, level, fmt, ...) macros.

Signed-off-by: Pisit Sawangvonganan <pisit@ndrsolution.com>
